tests/wscat_async.ml

```diff
@@ -33,55 +33,62 @@ let client uri =
   in
   Tcp.(with_connection (to_host_and_port host port) tcp_fun)
 
-(* let server uri = *)
-(*   let echo_fun id req recv send = *)
-(*     let open Frame in *)
-(*     let react fr = *)
-(*       match fr.opcode with *)
-(*       | Opcode.Ping -> *)
-(*         send @@ Frame.create ~opcode:Opcode.Pong ~content:fr.content () *)
-
-(*       | Opcode.Close -> *)
-(*         (\* Immediately echo and pass this last message to the user *\) *)
-(*         (if String.length fr.content >= 2 then *)
-(*            send @@ Frame.create ~opcode:Opcode.Close *)
-(*              ~content:(String.sub fr.content 0 2) () *)
-(*          else send @@ Frame.close 1000) >>= fun () -> *)
-(*         Lwt.fail Exit *)
-
-(*       | Opcode.Pong -> Lwt.return_unit *)
-
-(*       | Opcode.Text *)
-(*       | Opcode.Binary -> send fr *)
-
-(*       | _ -> *)
-(*         send @@ Frame.close 1002 >>= fun () -> *)
-(*         Lwt.fail Exit *)
-(*     in *)
-(*     let rec react_forever () = recv () >>= react >>= react_forever *)
-(*     in *)
-(*     react_forever () *)
-(*   in *)
-(*   Resolver_lwt.resolve_uri ~uri Resolver_lwt_unix.system >>= fun endp -> *)
-(*   Conduit_lwt_unix.(endp_to_server ~ctx:default_ctx endp >>= fun server -> *)
-(*   establish_server ~ctx:default_ctx ~mode:server echo_fun) *)
+let server uri =
+  let port = Option.value_exn ~message:"no port inferred from scheme"
+      Uri_services.(tcp_port_of_uri uri) in
+  Tcp.Server.create (Tcp.on_port port) begin fun addr reader writer ->
+    let app_to_ws, sender_write = Pipe.create () in
+    let receiver_read, ws_to_app = Pipe.create () in
+    let inet_addr = match addr with
+      | `Inet _ as a -> a
+    in
+    let finished = Websocket_async.server ~app_to_ws ~ws_to_app ~reader ~writer inet_addr in
+    let send = Pipe.write sender_write in
+    let rec loop () =
+      Pipe.read receiver_read >>= function
+      | `Eof -> Deferred.unit
+      | `Ok fr ->
+        begin
+          let open Frame in
+          match fr.opcode with
+          | Opcode.Ping ->
+              send @@
+              Frame.create ~opcode:Opcode.Pong ~content:fr.content () >>= fun () ->
+              (* Immediately echo and pass this last message to the user *)
+              if String.length fr.content >= 2 then
+                send @@ Frame.create ~opcode:Opcode.Close
```

Comment:
Is this correct? I ported this code from the lwt test, but it causes the connection to be closed when using this example with the client.